Mapper supports video streaming data reporting

wbc6080 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/kubeedge/kubeedge/blob/master/CONTRIBUTING.md
2. Ensure you have added or ran the appropriate tests for your PR

-->

**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind test
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind failing-test
-->


**What this PR does / why we need it**:

This PR implements mapper’s feature of supporting reporting of streaming data. Mainly includes the following parts:

1. Change the device model api definition to allow the data type of streaming data

2. Add the function of saving frame and saving video to the mapper-framework data processing function.

3. Change the logic of pushing device data in mapper-framework. When the type is streaming data, the device data is no longer pushed to the database and user applications, but processed by saving frames or saving videos.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes ##5401

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

wbc6080 commented:
/kind feature

wbc6080 commented:
PTAL,thanks @luomengY @WillardHu @fisherxu

luomengY commented:
Suggest adding a switch control parameter for starting/stopping stream data collection in the next version.

WillardHu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/handler.go:
```diff
@@ -0,0 +1,70 @@
+	visitorConfigData, err := json.Marshal(visitorConfig.VisitorConfigData)
+	err = json.Unmarshal(visitorConfigData, &streamConfig)
+	if err != nil {
+		klog.Errorf("Unmarshal streamConfigs error: %v", err)
+		return err
```
Wrap the error message and return it, make troubleshooting easy.
```golang
return fmt.Errorf("unmarshal streamConfigs error: %v", err)
```

wbc6080 replied:
Done

WillardHu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/img.go:
```diff
@@ -0,0 +1,246 @@
+	return fmt.Sprintf("%s/f%s.%s", dir, time.Now().Format(time.RFC3339Nano), format)
+}
+
+func save(frame *avutil.Frame, width int, height int, dir string, format string) {
```
Does this function not return an error?

wbc6080 replied:
I have add it

WillardHu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/img.go:
```diff
@@ -0,0 +1,246 @@
+		}
+	}
+	if streamIndex == -1 {
+		return fmt.Errorf("couldn't find video stream")
```
You can use the function errors.New() when without formatting.

WillardHu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/img.go:
```diff
@@ -0,0 +1,246 @@
+	// Assign appropriate parts of buffer to image planes in pFrameRGB
+	// Note that pFrameRGB is an AVFrame, but AVFrame is a superset
+	// of AVPicture
+	avp := (*avcodec.Picture)(unsafe.Pointer(pFrameRGB))
```
Should we consider the case where type is invalid?

wbc6080 replied:
I don't quite understand your question. Is there a problem with converting *avutil.Frame to *avcodec.Picture? The official documentation here shows that *avcodec.Picture is a superset of *avutil.Frame. You can refer to [this](https://github.com/giorgisio/goav/blob/master/example/tutorial01.go#L128)

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/README.md:
````diff
@@ -42,6 +42,30 @@ mapper
 ├── pkg ------------------------ Mapper register process, almost need not change
 └── Makefile
 ```
+In most cases, you need to fill in the code in the driver folder as follows: In the devicetype.go file, 
+you need to fill in the ProtocolConfig and VisitorConfig structure information in the way you defined in the instance yaml file, 
+so that the mapper can correctly parse the configuration information. In the driver.go file, 
+you need to customize the method of initializing the device and obtaining device data, and standardize 
+the data collected by the mapper. In config.yaml, you need to define the protocol name of mapper.
+## 2. Generate the mapper project
+After generating the mapper project and filling driver folder, users can make their own mapper image 
+based on the Dockerfile file and deploy the mapper in the cluster through deployment and other methods.
+```shell
+  docker build -t [YOUR MAPPER IMAGE NAME] .
+```
+If you want to debug locally first, you can also compile and run the mapper code directly:
+```shell
+  go run -tags Nostream cmd/main.go --v <log level,like 3> --config-file <path to config yaml>
+```
 
+The above operations are mainly for mappers that do not process streaming data. If you need to enable 
+the streaming data processing function, please execute the following command:
+```shell
+  go run -tags stream cmd/main.go --v <log level,like 3> --config-file <path to config yaml>
+```
````
It is not recommended to write too much here, just write it in the document.

wbc6080 replied:
I've simplified the README file but still kept some necessary steps to use it

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/README.md:
```diff
@@ -42,6 +42,30 @@ mapper
-mapper-framework is synced from https://github.com/kubeedge/kubeedge/tree/master/staging/src/github.com/kubeedge/mapper-framework. Code changes are made in that location, merged into kubeedge and later synced here.
+mapper-framework is synced from https://github.com/kubeedge/kubeedge/tree/master/staging/src/github.com/kubeedge/mapper-framework.
+Code changes are made in that location, merged into kubeedge and later synced here.
+# Things you should *NOT* do
+https://github.com/kubeedge/kubeedge/tree/master/staging/src/github.com/kubeedge/mapper-framework is synced to kubeedge/mapper-framework. 
+All changes must be made in the former. The latter is read-only.
```
It shoud not need this, and keep consistent with other sub-projects :)

wbc6080 replied:
Done

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/Dockerfile:
```diff
@@ -7,7 +7,7 @@ ENV GO111MODULE=on \
 COPY . .
 
-RUN CGO_ENABLED=0 GOOS=linux go build -o main cmd/main.go
+RUN CGO_ENABLED=0 GOOS=linux go build -tags Nostream -o main cmd/main.go
```
Is there any way to control the feature? Like add a startup parameter to control the mdule to start or not, hdyt @WillardHu @luomengY @wbc6080

wbc6080 replied:
I have considered using a switch to set whether to use this feature, but even if we add a switch, it seems that this part of the code will still be compiled, only not run this part, so the third-party package of stream data processing will still be used. In order to prevent this part of the code from being compiled, what I finally found was the method of conditional compilation. As for other methods, I haven't found them yet.

WillardHu replied:
One idea is that handle the logic(determine which files to copy) when the mapper is generated from the template.

wbc6080 replied:
Oh I see, very good idea, I will try to implement it like this @WillardHu

wbc6080 commented:
> Suggest adding a switch control parameter for starting/stopping stream data collection in the next version.

Yes, we can add it with the feature about mapper supporting data writing in v1.18

wbc6080 commented:
I have made modifications according to everyone's opinions, mainly:
1. Removed conditional compilation.
2. Modified the mapper generate script to generate different mapper projects depending on whether it is used to process stream data.
3. Provide Dockerfile files for two situations
PTAL,thanks @WillardHu @fisherxu

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/handler_nostream.go:
```diff
@@ -0,0 +1,33 @@
+/*
```
Is this file necessary? The file `handler_nostream.go` under `stream` dir seems a bit unreasonable

wbc6080 replied:
This file is required, because stream.Handler needs to be called at mapper data plane capability like picture following. For mappers that do not require stream data processing, if this file is removed, an error will occur during compilation because stream.Handler cannot be found. If for name reasons, I can modify the mapper generation script and name this file to handler.go in the generated mapper project.

![device](https://github.com/kubeedge/kubeedge/assets/48212997/1b36ab8c-5222-4257-bc56-5a01c826b704)

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/handler.go:
```diff
@@ -0,0 +1,66 @@
+	default:
+		err = fmt.Errorf("cannot find the processing method for the corresponding Property %s of the stream data", twin.PropertyName)
+	}
+	if err == nil {
+		klog.V(2).Infof("Successfully processed streaming data by %s", twin.PropertyName)
+	}
+	return err
```
```suggestion
	if err != nil {
	        return err
	}
	klog.V(2).Infof("Successfully processed streaming data by %s", twin.PropertyName)
	return nil
```

wbc6080 replied:
Done

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/handler.go:
```diff
@@ -0,0 +1,66 @@
+	switch twin.PropertyName {
+	// Currently, the function of saving frames and saving videos is built-in according to the configuration.
+	// Other functions can be expanded here.
+	case "saveFrame":
+		err = SaveFrame(streamURI.(string), streamConfig.OutputDir, streamConfig.Format, streamConfig.FrameCount, streamConfig.FrameInterval)
+	case "saveVideo":
```
The PropertyName name is passed from Device API? If so, It's necessary to add some comments in api and add some const

wbc6080 replied:
Done

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/_template/mapper/Dockerfile_stream:
```diff
@@ -0,0 +1,18 @@
+    ./configure &&  make && \
+    make install
+
+RUN GOOS=linux go build -o main cmd/main.go
```
Is it necessary to separate into two dockerfile? also cc @WillardHu

wbc6080 replied:
Since the stream mapper needs to download additional dependencies, it takes about five minutes to build the image, so I feel it would be better to separate it without affecting the construction of other mapper images.

fisherxu commented on staging/src/github.com/kubeedge/mapper-framework/hack/make-rules/generate.sh:
```diff
@@ -27,6 +33,13 @@ function entry() {
+  fi
+
+  if [ "${buildMethod}" = "nostream" ]; then
+      rm "${mapperPath}/data/stream/handler.go" "${mapperPath}/data/stream/img.go" "${mapperPath}/data/stream/video.go"
+  fi
```
If add new file for stream, this part will be hard to maintain.

wbc6080 replied:
I modified it to only keep the handler_stream.go file and renamed this file to handler.go after build

fisherxu reviewed (commented):
/approve
Looks good for the first version.
/assign @WillardHu

WillardHu commented:
/lgtm
